File: Ilya-Dudnikov/term_2/block_3/task_2/App/src/main/java/view/WeatherView.java
package view;

import controller.StateController;
import model.WeatherData.WeatherData;

public class WeatherView {
	private WeatherData data;

	public WeatherView(WeatherData data) {
		this.data = data;
	}

	private String valueOrMessage(Double val) {
		if (val == null)
			return "Data unavailable :(";
		return val.toString();
	}

	public void outputData() {
		System.out.println(
				"Air Temperature, C\u00B0: " + valueOrMessage(data.airTemperatureC()) + System.lineSeparator()
				+ "Air Temperature, F\u00B0: " + valueOrMessage(data.airTemperatureF()) + System.lineSeparator()
				+ "Cloud Cover, %: " + valueOrMessage(data.cloudCover()) + System.lineSeparator()
				+ "Humidity, %: " + valueOrMessage(data.humidity()) + System.lineSeparator()
				+ "Precipitation, mm/h: " + valueOrMessage(data.precipitation()) + System.lineSeparator()
				+ "Wind Direction, \u00B0: " + valueOrMessage(data.windDirection()) + System.lineSeparator()
				+ "Wind Speed, m/s: " + valueOrMessage(data.windSpeed())
		);
	}
}

File: IvanShurenkov/Second Term/Block4/Task6/app-tornadofx/src/main/kotlin/UI/WeatherView.kt
package UI

import KodeinInjactions
import javafx.beans.property.SimpleStringProperty
import javafx.collections.ObservableList
import lib.weather.IWeatherApi
import lib.weather.date.Location
import lib.weather.date.Weather
import org.kodein.di.instance
import tornadofx.*
import kotlin.concurrent.thread

class WeatherView : View() {
    private var services: ObservableList<String> = observableListOf()
    private val selectedService = SimpleStringProperty()

    private var tempreture = SimpleStringProperty()
    private var cloudCoverage = SimpleStringProperty()
    private var humidity = SimpleStringProperty()
    private var precipitation = SimpleStringProperty()
    private var windSpeed = SimpleStringProperty()
    private var windDirection = SimpleStringProperty()

    private val noData = "No data available"
    private val hasNotService = "Service had been not selected"

    init {
        for (i in KodeinInjactions.servicesNames) {
            services += i
        }
        tempreture.value = hasNotService
        cloudCoverage.value = hasNotService
        humidity.value = hasNotService
        precipitation.value = hasNotService
        windSpeed.value = hasNotService
        windDirection.value = hasNotService
    }

    override val root = vbox {
        combobox(selectedService, services) {
            setMinSize(300.0, 10.0)
            setOnAction {
                println(selectedService.value)
                val service: IWeatherApi by KodeinInjactions.service.instance(selectedService.value)
                updateLabel(service.getWeather())
            }
        }
        hbox {
            vbox {
                setMinSize(110.0, 0.0)
                label("Temperature")
                label("Cloud coverage")
                label("Humidity")
                label("Precipitation")
                label("Wind speed")
                label("Wind directiom")
            }
            vbox {
                label(tempreture)
                label(cloudCoverage)
                label(humidity)
                label(precipitation)
                label(windSpeed)
                label(windDirection)
            }
        }
        button("Update") {
            setMinSize(300.0, 10.0)
            action {
                if (selectedService.value != null) {
                    val service: IWeatherApi by KodeinInjactions.service.instance(selectedService.value)
                    val location: Location by KodeinInjactions.location.instance("spb")
                    val api: String by KodeinInjactions.apikey.instance(selectedService.value)
                    if (api != null)
                        service.updateWeather(location, api)
                    updateLabel(service.getWeather())
                }
            }
        }
    }

    fun updateLabel(weather: Weather) {
        tempreture.value =
            if (weather.temperature != null) "${weather.temperature!!.celsius}°C | ${weather.temperature!!.fahrenheit}°F"
            else noData
        cloudCoverage.value = if (weather.cloudCoverage != null) "${weather.cloudCoverage!!.percent}%"
        else noData
        humidity.value = if (weather.humidity != null) "${weather.humidity!!.percent}%"
        else noData
        precipitation.value = if (weather.precipitation != null) "${weather.precipitation!!.mmPerHour} mm/h"
        else noData
        windSpeed.value = if (weather.windSpeed != null) "${weather.windSpeed!!.speed} m/s"
        else noData
        windDirection.value = if (weather.windDirection != null) "${weather.windDirection!!.degree}°"
        else noData
    }
}
